### PR TITLE
tabula-java allow specifying multiple areas

### PR DESCRIPTION
**Pull request by asheeshrana**
1942-update-tabula-java-for-muti-column-pdf (PolicyReporter/requests/#1942)
- Allow multiple occurrences of -a parameter
- Allow -a parameter to accept % values as well as absolute values
- Add few test cases
- Add test file

Misc Notes for reviewer:
- Added a new class Pair. Created only a very basic implementation. Could be replaced in future by [apache-commons-lang](https://commons.apache.org/proper/commons-lang/apidocs/org/apache/commons/lang3/tuple/Pair.html)
- Areas are stored in list of "Pairs" so that tables are extracted in the same order as specified in the commandline
- Haven't put any limits on values for % option as specifying length/width which is bigger than the pdf page dimensions is allowed even now with absolute co-ordinates.

---
This change is [Reviewable](https://reviewable.io/reviews/tabulapdf/tabula-java/213)

### CODE REVIEW

File: src/main/java/technology/tabula/Page.java [hunk 1]
```diff
@@ -48,6 +51,21 @@ public Page(float top, float left, float width, float height, int rotation, int
     }
 
     
+    public Page getArea(float top, float left, float bottom, float right, int mode) {
+        Rectangle area = new Rectangle(top, left, right - left, bottom - top);
+        return getArea(area, mode);
+    }
+
+    public Page getArea(Rectangle area, int mode) {
```
Comment (jazzido) [inline src/main/java/technology/tabula/Page.java]: I'd rather leave this method as it was, and make the _relative_→_absolute_ conversion in `CommandLineApp`.

File: src/main/java/technology/tabula/CommandLineApp.java [hunk 2]
```diff
@@ -156,11 +160,13 @@ private void extractFile(File pdfFile, Appendable outFile) throws ParseException
             while (pageIterator.hasNext()) {
                 Page page = pageIterator.next();
 
-                if (pageArea != null) {
-                    page = page.getArea(pageArea);
+                if (pageAreas != null) {
+                    for (Pair<Integer, Rectangle> areaPair : pageAreas) {
```
Comment (jazzido) [inline src/main/java/technology/tabula/CommandLineApp.java]: I mean here. You have access to the `Page` objects, so you can do the conversion at this point.

**Comment by jazzido** [comment]
This is awesome, @asheeshrana — Thanks a lot! The tests are very much appreciated :)

I've requested a couple minor changes, let me know what you think.

File: src/main/java/technology/tabula/Page.java (hunk 1 above)
Comment (asheeshrana) [reply src/main/java/technology/tabula/Page.java]: The orginal method is there as is... I didn't modify it. I simply added another method.
Additionally based on the comment at the top of the file...
~~~
// TODO: this class should probably be called "PageArea" or something like that
~~~
it seemed to me that Page.java is the place where all area calculations should be housed and hence I decided the add that method.

I can anyways move the calculations out of Page.java... but just want a quick confirmation from you again if possible please.

File: src/main/java/technology/tabula/CommandLineApp.java (hunk 2 above)
Comment (asheeshrana) [reply src/main/java/technology/tabula/CommandLineApp.java]: Agree... tried to explain the thinking behind why I put the calculations the Page.java as opposed to CommandLineApp.java file... Please do confirm once more before I send the updated code.

PS: I apologize for asking to reconfirm

Comment (asheeshrana) [reply src/main/java/technology/tabula/CommandLineApp.java]: Explanation is under the other review comment.

File: src/main/java/technology/tabula/Page.java (hunk 1 above)
Comment (jazzido) [reply src/main/java/technology/tabula/Page.java]: Sorry about that. I misread and thought you modified the original method. My bad.

In any case, I still feel that the conversion belongs in `CommandLineApp`.

File: src/main/java/technology/tabula/CommandLineApp.java (hunk 2 above)
Comment (jazzido) [reply src/main/java/technology/tabula/CommandLineApp.java]: No apologies needed!

Comment (asheeshrana) [reply src/main/java/technology/tabula/CommandLineApp.java]: Changes done and committed.

**Comment by jazzido** [comment]
Merged! Thanks a lot, @asheeshrana !

**Pull request merged**